File: modules/core/src/encoder/segment/OOBVendorsEncoder.ts
import {

  SegmentType,

} from '.';

import {

  BitLength,
  Base64Url,

} from '../';

import {

  FieldEncoderMap,
  IntEncoder,

} from '../field';

import {

  EncodingError,

} from '../../errors';

import {

  TCModel,

} from '../../';

import {

  Vector,

} from '../../model';

export class OOBVendorsEncoder {

  private base64Url: Base64Url = new Base64Url();

  public static encode(tcModel: TCModel, type: string ): string {

    const encMap: FieldEncoderMap = new FieldEncoderMap();

    // if we have an encoder, we can do it!
    if (!encMap[type]) {

      throw new EncodingError(`invalid type: ${type}`);

    }

    const encoder = encMap[type];

    // first encode the segment type
    let bits: string = IntEncoder.encode(SegmentType[type], BitLength.segmentType);

    // add the vector bits
    bits += encoder.encode(tcModel[type]);

    return Base64Url.encode(bits);

  }

  public static decode(encodedString: string, tcModel: TCModel): TCModel {

    const encMap: FieldEncoderMap = new FieldEncoderMap();

    // first get this into bits
    const bits: string = Base64Url.decode(encodedString);

    /**
     * get the segType bits of the front of the bitfield decode them into a
     * number
     */
    const segType: number = (encMap.segmentType).decode(bits.substr(0, BitLength.segmentType));
    const segmentName: string = SegmentType[segType.toString()];
    const encoder = encMap[segmentName];
    // less the segmentType bits
    const vector: Vector = encoder.decode(bits.substr(BitLength.segmentType)) as Vector;

    vector.forEach((isSet: boolean, vendorId: number): void => {

      if (isSet) {

        tcModel[segmentName].set(vendorId);

      }

    });

    return tcModel;

  }

}

File: GDPR-Phase2-Java/src/main/java/com/gdpr/phase2/encoder/segment/OOBVendorsEncoder.java
package com.gdpr.phase2.encoder.segment;

import com.gdpr.phase2.encoder.Base64Url;
import com.gdpr.phase2.model.SortedVector;
import com.gdpr.phase2.Utils.TCModel;
import com.gdpr.phase2.Utils.TCModelEnum;
import com.gdpr.phase2.encoder.BaseEncoder;
import com.gdpr.phase2.encoder.BitLength;
import com.gdpr.phase2.encoder.field.FieldEncoderMap;

import java.util.Map;
import java.util.Optional;


public class OOBVendorsEncoder implements BaseSegmentEncoder {
    private OOBVendorsEncoder() {
    }
    private static final OOBVendorsEncoder instance = new OOBVendorsEncoder();
    public static OOBVendorsEncoder getInstance() {
        return instance;
    }

    public TCModel decode(String encodedString, TCModel tcModel) {
        final Map<String, BaseEncoder> encMap = FieldEncoderMap.getInstance().getFieldMap();
        String bits = Base64Url.decode(encodedString);
        int index = 0;
        int segType = (int)(encMap.get("segmentType").decode(bits.substring(index,index + Optional.ofNullable(BitLength.fieldLengths.get("segmentType")).orElse(0))));
        String segmentName = "";
        switch(segType) {
            case 0:
                segmentName = SegmentType.zero;
                break;
            case 1:
                segmentName = SegmentType.one;
                break;
            case 2:
                segmentName = SegmentType.two;
                break;
            case 3:
                segmentName = SegmentType.three;
                break;
        }
        BaseEncoder<SortedVector> encoder = (BaseEncoder<SortedVector>) encMap.get(segmentName);
        SortedVector vector = encoder.decode(bits.substring(Optional.ofNullable(BitLength.fieldLengths.get("segmentType")).orElse(0)));

        SortedVector set = new SortedVector();
        for(int i=0; i <= (int)vector.getSet().first(); i++) {
            if(vector.getSet().contains(i)){
                set = (SortedVector) TCModelEnum.valueOf(segmentName).getValue(tcModel);
                set.getSet().add(i);
                if(set.getBitLength()!=0) {
                    set.setBitLength(0);
                }

            }

        }
        if(set.getSet().size()>0) {
            TCModelEnum.valueOf(segmentName).setValue(tcModel, set);
        }
        return tcModel;



    }


}
